File: src/com/verilang/psi/factory/nodes/HierarchicalIdentifierPsiNode.java
package com.verilang.psi.factory.nodes;

import com.intellij.lang.ASTNode;
import com.intellij.psi.PsiReference;
import com.intellij.psi.util.PsiTreeUtil;
import org.antlr.jetbrains.adaptor.psi.ANTLRPsiNode;
import org.jetbrains.annotations.NotNull;


public class HierarchicalIdentifierPsiNode extends ANTLRPsiNode {

    public HierarchicalIdentifierPsiNode(@NotNull ASTNode node) {
        super(node);
    }

    @NotNull
    @Override
    public PsiReference[] getReferences() {
        return PsiTreeUtil.getChildrenOfTypeAsList(
                this.getNode().getPsi(),
                SimpleIdentifierPsiLeafNode.class
        ).stream()
                .map(SimpleIdentifierPsiLeafNode::getReference)
                .toArray(PsiReference[]::new);
    }
}

File: src/com/verilang/psi/factory/nodes/HierarchicalIdentifierPsiNode.kt
package com.verilang.psi.factory.nodes

import com.intellij.lang.ASTNode
import com.intellij.openapi.util.TextRange
import com.intellij.psi.PsiReference
import com.intellij.psi.util.PsiTreeUtil
import com.verilang.psi.references.HierarchicalIdentifierReference
import org.antlr.jetbrains.adaptor.psi.ANTLRPsiNode


class HierarchicalIdentifierPsiNode(node: ASTNode)
    : ANTLRPsiNode(node), ReferenceHolder<SimpleIdentifierPsiLeafNode> {

    override fun getName(): String? {
        return this.getHoldPsiNode()?.name
    }

    public override fun getHoldPsiNode(): SimpleIdentifierPsiLeafNode? {
        return PsiTreeUtil
                .findChildOfType(this, SimpleIdentifierPsiLeafNode::class.java)
    }

    public override fun getHoldPsiNodeRelativeTextRange(): TextRange? {
        return getHoldPsiNode()
                ?.textRange
                ?.shiftLeft(this.textOffset)
    }

    override fun getReference(): PsiReference? {
        return HierarchicalIdentifierReference(this)
    }

}
